payment_methods/Paypal_Express/EEG_Paypal_Express.gateway.php

```diff
@@ -126,490 +130,169 @@ public function set_settings($settings_array)
             'FHEE__EEG_Paypal_Express__set_redirection_info__arguments',
-            $token_request_dtls,
+            $this->token_request->generateDetails(
+                $payment,
+                $this->getOrderItems($payment),
+                $return_url ?? '',
+                $cancel_url ?? '',
+                $this->_image_url ?? '',
+                $this->_request_shipping_addr ?? ''
+            ),
             $this
         );
         // Request PayPal token.
-        $token_request_response = $this->_ppExpress_request($token_request_dtls, 'Payment Token', $payment);
-        $token_rstatus = $this->_ppExpress_check_response($token_request_response);
-        $response_args = (isset($token_rstatus['args']) && is_array($token_rstatus['args']))
-            ? $token_rstatus['args']
-            : array();
-        if ($token_rstatus['status']) {
-            // We got the Token so we may continue with the payment and redirect the client.
-            $payment->set_details($response_args);
-            $gateway_url = $this->_debug_mode ? 'https://www.sandbox.paypal.com' : 'https://www.paypal.com';
-            $payment->set_redirect_url(
-                $gateway_url
-                . '/checkoutnow?useraction=commit&cmd=_express-checkout&token='
-                . $response_args['TOKEN']
-            );
-        } else {
-            if (isset($response_args['L_ERRORCODE'])) {
-                $payment->set_gateway_response($response_args['L_ERRORCODE'] . '; ' . $response_args['L_SHORTMESSAGE']);
-            } else {
-                $payment->set_gateway_response(
-                    esc_html__(
-                        'Error occurred while trying to setup the Express Checkout.',
-                        'event_espresso'
-                    )
-                );
-            }
-            $payment->set_details($response_args);
-            $payment->set_status($this->_pay_model->failed_status());
-        }
+        $token_request_response = $this->_ppExpress_request($token_request_details, 'Payment Token', $payment);
+        $token_request_status   = $this->_ppExpress_check_response($token_request_response);
+        $this->token_request->processResponse($payment, $token_request_status, $this->isInSandboxMode());
+
         return $payment;
     }
 
 
-
     /**
      * @param array           $update_info {
      * @type string           $gateway_txn_id
-     * @type string status an EEMI_Payment status
+     * @type string           $status      an EEMI_Payment status
      *                                     }
      * @param EEI_Transaction $transaction
      * @return EEI_Payment
+     * @throws Exception
      */
-    public function handle_payment_update($update_info, $transaction)
+    public function handle_payment_update($update_info, $transaction): EEI_Payment
     {
-        $payment = $transaction instanceof EEI_Transaction ? $transaction->last_payment() : null;
```

Comment:
i hate trying to debug functions that are hundreds of lines long so I extracted some of the PayPal gateway logic into new classes with more sensible function sizes. SRP FTW!